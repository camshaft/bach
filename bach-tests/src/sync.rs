use bach::{
    ext::*,
    sync::{Mutex, Semaphore},
};
use std::sync::Arc;

mod mpsc;
mod mutex;
mod rwlock;
mod semaphore;

#[cfg(test)]
pub fn sim(f: impl Fn()) -> impl Fn() {
    crate::testing::init_tracing();
    move || {
        let mut rt = bach::environment::default::Runtime::new()
            .with_coop(true)
            .with_rand(None);
        rt.run(&f);
    }
}
