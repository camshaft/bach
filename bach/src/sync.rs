pub mod channel;
pub mod queue;
