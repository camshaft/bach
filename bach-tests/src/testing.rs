use std::{collections::HashMap, sync::Mutex};

pub fn init_tracing() {
    if !cfg!(test) || cfg!(feature = "leaks") {
        return;
    }

    use std::sync::Once;

    static TRACING: Once = Once::new();

    // make sure this only gets initialized once
    TRACING.call_once(|| {
        let format = tracing_subscriber::fmt::format()
            .with_level(false) // don't include levels in formatted output
            .with_timer(Uptime)
            .with_ansi(false)
            .compact(); // Use a less verbose output format.

        struct Uptime;

        // Generate the timestamp from the testing IO provider rather than wall clock.
        impl tracing_subscriber::fmt::time::FormatTime for Uptime {
            fn format_time(
                &self,
                w: &mut tracing_subscriber::fmt::format::Writer<'_>,
            ) -> std::fmt::Result {
                let now =
                    bach::time::scheduler::scope::try_borrow_mut_with(|s| Some(s.as_ref()?.now()));
                if let Some(now) = now {
                    write!(w, "{now}")
                } else {
                    write!(w, "[UNKNOWN]")
                }
            }
        }

        let env_filter = tracing_subscriber::EnvFilter::builder()
            .with_default_directive(tracing::Level::DEBUG.into())
            .with_env_var("BACH_LOG")
            .from_env()
            .unwrap();

        tracing_subscriber::fmt()
            .with_env_filter(env_filter)
            .event_format(format)
            .with_test_writer()
            .init();
    });
}

pub trait Event: Clone + core::fmt::Debug + Eq + core::hash::Hash {
    fn is_start(&self) -> bool;
}

pub struct Log<T>(Mutex<Vec<T>>);

impl<T: Event> Default for Log<T> {
    fn default() -> Self {
        Self::new()
    }
}

impl<T: Event> Log<T> {
    pub const fn new() -> Self {
        Self(Mutex::new(Vec::new()))
    }

    pub fn push(&self, event: T) {
        self.0.lock().unwrap().push(event);
    }

    pub fn check(&self) -> Vec<T> {
        let mut sequences = vec![];
        let mut current_sequence = vec![];
        let mut seen = HashMap::<_, usize>::new();

        let events = core::mem::take(&mut *self.0.lock().unwrap());

        for event in events.iter() {
            if event.is_start() {
                if !current_sequence.is_empty() {
                    let seq = std::mem::take(&mut current_sequence);
                    *seen.entry(seq.clone()).or_default() += 1;
                    sequences.push(seq);
                }
            } else {
                current_sequence.push(event.clone());
            }
        }

        if !current_sequence.is_empty() {
            let seq = std::mem::take(&mut current_sequence);
            *seen.entry(seq.clone()).or_default() += 1;
            sequences.push(seq);
        }

        let mut duplicate = false;

        for (seq, count) in seen {
            if count == 1 {
                continue;
            }
            duplicate = true;
            eprintln!("duplicate sequence ({count}): {seq:#?}");
        }

        assert!(!duplicate, "duplicate interleavings found");
        assert!(!sequences.is_empty(), "no event sequences recorded");

        eprintln!("Found {} unique event sequences", sequences.len());

        events
    }
}
