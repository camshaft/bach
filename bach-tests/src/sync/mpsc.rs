use super::*;
use crate::testing::Log;
use bach::sync::mpsc::{channel, unbounded_channel};

#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
enum Event {
    Start,
    SendStarted { task: usize },
    SendCompleted { task: usize },
    RecvStarted { task: usize },
    RecvCompleted { task: usize, value: Option<i32> },
}

impl crate::testing::Event for Event {
    fn is_start(&self) -> bool {
        matches!(self, Event::Start)
    }
}

#[test]
fn mpsc_try_operations() {
    bolero::check!().exhaustive().run(sim(|| {
        async {
            let (tx, mut rx) = channel(2);

            // Try sending messages
            tx.try_send(1).unwrap();
            tx.try_send(2).unwrap();

            // Channel should be full now
            match tx.try_send(3) {
                Ok(_) => panic!("Channel should be full"),
                Err(e) => match e {
                    bach::sync::mpsc::error::TrySendError::Full(_) => {} // expected
                    _ => panic!("Unexpected error: {e:?}"),
                },
            }

            // Try receiving messages
            assert_eq!(rx.try_recv().unwrap(), 1);
            assert_eq!(rx.try_recv().unwrap(), 2);

            // Channel should be empty now
            assert!(rx.try_recv().is_err());
        }
        .primary()
        .spawn();
    }));
}

#[test]
fn mpsc_send_recv() {
    static LOG: Log<Event> = Log::new();

    bolero::check!().exhaustive().run(sim(|| {
        LOG.push(Event::Start);

        let (tx, mut rx) = channel(1);

        // Spawn a task that sends two values (will block on the second one)
        async move {
            LOG.push(Event::SendStarted { task: 1 });
            tx.send(42).await.unwrap();
            LOG.push(Event::SendCompleted { task: 1 });

            LOG.push(Event::SendStarted { task: 1 });
            tx.send(43).await.unwrap();
            LOG.push(Event::SendCompleted { task: 1 });
        }
        .primary()
        .spawn();

        // Spawn a task that receives values
        async move {
            LOG.push(Event::RecvStarted { task: 2 });
            let val = rx.recv().await;
            LOG.push(Event::RecvCompleted {
                task: 2,
                value: val,
            });

            LOG.push(Event::RecvStarted { task: 2 });
            let val = rx.recv().await;
            LOG.push(Event::RecvCompleted {
                task: 2,
                value: val,
            });
        }
        .primary()
        .spawn();
    }));

    insta::assert_debug_snapshot!(LOG.check());
}

#[test]
fn mpsc_multiple_senders() {
    static LOG: Log<Event> = Log::new();

    bolero::check!().exhaustive().run(sim(|| {
        LOG.push(Event::Start);

        let (tx, mut rx) = channel(2);

        // First sender task
        for task in 1..=2 {
            let tx = tx.clone();
            async move {
                LOG.push(Event::SendStarted { task });
                tx.send((task * 10) as _).await.unwrap();
                LOG.push(Event::SendCompleted { task });
            }
            .primary()
            .spawn_named(format!("Sender {task}"));
        }

        // Receiver task
        async move {
            LOG.push(Event::RecvStarted { task: 3 });
            let val = rx.recv().await;
            LOG.push(Event::RecvCompleted {
                task: 3,
                value: val,
            });

            LOG.push(Event::RecvStarted { task: 3 });
            let val = rx.recv().await;
            LOG.push(Event::RecvCompleted {
                task: 3,
                value: val,
            });
        }
        .primary()
        .spawn_named("Receiver");
    }));

    insta::assert_debug_snapshot!(LOG.check());
}

#[test]
fn mpsc_channel_closing() {
    static LOG: Log<Event> = Log::new();

    bolero::check!().exhaustive().run(sim(|| {
        LOG.push(Event::Start);

        let (tx, mut rx) = channel::<i32>(1);

        // Task to close the sender
        async move {
            // Let's send one message first
            LOG.push(Event::SendStarted { task: 1 });
            tx.send(100).await.unwrap();
            LOG.push(Event::SendCompleted { task: 1 });

            // Close by dropping
            drop(tx);
        }
        .primary()
        .spawn_named("Sender");

        // Receiver task that continues after channel is closed
        async move {
            // First message should come through
            LOG.push(Event::RecvStarted { task: 2 });
            let val = rx.recv().await;
            LOG.push(Event::RecvCompleted {
                task: 2,
                value: val,
            });

            // Second receive should get None (channel closed)
            LOG.push(Event::RecvStarted { task: 2 });
            let val = rx.recv().await;
            LOG.push(Event::RecvCompleted {
                task: 2,
                value: val,
            });
        }
        .primary()
        .spawn_named("Receiver");
    }));

    insta::assert_debug_snapshot!(LOG.check());
}

#[test]
fn mpsc_permit_operations() {
    static LOG: Log<Event> = Log::new();

    bolero::check!().exhaustive().run(sim(|| {
        LOG.push(Event::Start);

        let (tx, mut rx) = channel(1);

        // Task to send using permits
        async move {
            LOG.push(Event::SendStarted { task: 1 });
            let permit = tx.reserve().await.unwrap();
            permit.send(50);
            LOG.push(Event::SendCompleted { task: 1 });
        }
        .primary()
        .spawn_named("Sender");

        // Receiver task
        async move {
            LOG.push(Event::RecvStarted { task: 2 });
            let val = rx.recv().await;
            LOG.push(Event::RecvCompleted {
                task: 2,
                value: val,
            });
        }
        .primary()
        .spawn_named("Receiver");
    }));

    insta::assert_debug_snapshot!(LOG.check());
}

#[test]
fn mpsc_unbounded_channel_operations() {
    static LOG: Log<Event> = Log::new();

    bolero::check!().exhaustive().run(sim(|| {
        LOG.push(Event::Start);

        let (tx, mut rx) = unbounded_channel::<i32>();

        // Task to send multiple messages without blocking
        async move {
            for i in 0..5 {
                LOG.push(Event::SendStarted { task: 1 });
                tx.send(i).unwrap();
                LOG.push(Event::SendCompleted { task: 1 });
            }
        }
        .primary()
        .spawn_named("Sender");

        // Receiver task
        async move {
            for _ in 0..5 {
                LOG.push(Event::RecvStarted { task: 2 });
                let val = rx.recv().await;
                LOG.push(Event::RecvCompleted {
                    task: 2,
                    value: val,
                });
            }
        }
        .primary()
        .spawn_named("Receiver");
    }));

    insta::assert_debug_snapshot!(LOG.check());
}

#[test]
fn mpsc_weak_sender_operations() {
    static LOG: Log<Event> = Log::new();

    bolero::check!().exhaustive().run(sim(|| {
        LOG.push(Event::Start);

        let (tx, mut rx) = channel::<i32>(1);

        // Create a weak sender but DON'T drop the original sender yet
        let weak_tx = tx.downgrade();

        // Task to upgrade and use weak sender
        async move {
            LOG.push(Event::SendStarted { task: 1 });
            if let Some(tx) = weak_tx.upgrade() {
                tx.send(75).await.unwrap();
                LOG.push(Event::SendCompleted { task: 1 });
            } else {
                panic!("Failed to upgrade weak sender");
            }

            // A sender that does not prevent the channel from being closed.
            //
            // If all Sender instances of a channel were dropped and only WeakSender
            // instances remain, the channel is closed.
            drop(tx);

            assert!(weak_tx.upgrade().is_none());
        }
        .primary()
        .spawn_named("Weak Sender");

        // Receiver task
        async move {
            LOG.push(Event::RecvStarted { task: 2 });
            let val = rx.recv().await;
            LOG.push(Event::RecvCompleted {
                task: 2,
                value: val,
            });
        }
        .primary()
        .spawn_named("Receiver");
    }));

    insta::assert_debug_snapshot!(LOG.check());
}

#[test]
fn mpsc_reserve_many() {
    static LOG: Log<Event> = Log::new();

    bolero::check!().exhaustive().run(sim(|| {
        LOG.push(Event::Start);

        let (tx, mut rx) = channel(3);

        // Task to reserve multiple slots
        async move {
            LOG.push(Event::SendStarted { task: 1 });
            let permits = tx.reserve_many(2).await.unwrap();

            let mut count = 0;
            for permit in permits {
                permit.send(count);
                count += 1;
            }
            LOG.push(Event::SendCompleted { task: 1 });
        }
        .primary()
        .spawn_named("Multi-Sender");

        // Receiver task
        async move {
            for _ in 0..2 {
                LOG.push(Event::RecvStarted { task: 2 });
                let val = rx.recv().await;
                LOG.push(Event::RecvCompleted {
                    task: 2,
                    value: val,
                });
            }
        }
        .primary()
        .spawn_named("Receiver");
    }));

    insta::assert_debug_snapshot!(LOG.check());
}
