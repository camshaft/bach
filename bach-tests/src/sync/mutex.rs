use super::*;
use crate::testing::Log;

#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
enum Event {
    Start,
    MutexAcquired { task: usize },
    MutexReleased { task: usize },
}

impl crate::testing::Event for Event {
    fn is_start(&self) -> bool {
        matches!(self, Event::Start)
    }
}

#[test]
fn mutex_try_lock() {
    bolero::check!().exhaustive().run(sim(|| {
        async {
            let mutex = Mutex::new(5);

            // Successfully try_lock
            let mut guard = mutex.try_lock().unwrap();
            *guard = 10;

            // This should fail while the first lock is held
            assert!(mutex.try_lock().is_err());

            // Release the first lock
            drop(guard);

            // Now try_lock should succeed
            let guard = mutex.try_lock().unwrap();
            assert_eq!(*guard, 10);
        }
        .primary()
        .spawn();
    }));
}

#[test]
fn mutex_guard_deref() {
    bolero::check!().exhaustive().run(sim(|| {
        async {
            let mutex = Mutex::new(String::from("hello"));

            // Test that we can access the value through the guard
            let guard = mutex.lock().await;
            assert_eq!(guard.len(), 5);
            assert_eq!(*guard, "hello");
            drop(guard);

            // Test that we can mutate the value through the guard
            let mut guard = mutex.lock().await;
            guard.push_str(" world");
            assert_eq!(*guard, "hello world");
        }
        .primary()
        .spawn();
    }));
}

#[test]
fn mutex_lock_contention() {
    // For tracking test events
    static LOG: Log<Event> = Log::new();

    bolero::check!().exhaustive().run(sim(|| {
        let mutex = Arc::new(Mutex::new(0));

        // Start a task that holds the lock for a while
        {
            let mutex = mutex.clone();
            async move {
                let mut guard = mutex.lock().await;
                *guard = 1;
                LOG.push(Event::MutexAcquired { task: 0 });

                // Hold the lock for some time
                bach::time::sleep(std::time::Duration::from_millis(20)).await;
                *guard = 2;

                drop(guard);
                LOG.push(Event::MutexReleased { task: 0 });
            }
            .primary()
            .spawn();
        }

        // Try to acquire a different mutex to avoid clone issues
        async move {
            // Acquire our own mutex
            let guard = mutex.lock().await;
            LOG.push(Event::MutexAcquired { task: 1 });

            drop(guard);
            LOG.push(Event::MutexReleased { task: 1 });
        }
        .primary()
        .spawn();
    }));

    insta::assert_debug_snapshot!(LOG.check());
}

#[test]
fn mutex_interleavings_snapshot() {
    static LOG: Log<Event> = Log::new();

    bolero::check!().exhaustive().run(sim(|| {
        // Start a new event sequence
        LOG.push(Event::Start);

        // Create a mutex
        let mutex = Arc::new(Mutex::new(0));

        // Launch multiple tasks that will contend for the mutex
        for task_id in 0..3 {
            let mutex = mutex.clone();
            async move {
                // Acquire the mutex
                let mut value = mutex.lock().await;
                LOG.push(Event::MutexAcquired { task: task_id });

                // Do some work with the mutex
                *value += 1;

                // Release the mutex (implicit when guard is dropped)
                drop(value);
                LOG.push(Event::MutexReleased { task: task_id });
            }
            .primary()
            .spawn_named(format!("Task {task_id}"));
        }
    }));

    insta::assert_debug_snapshot!(LOG.check());
}

#[test]
fn mutex_owned_lock() {
    bolero::check!().exhaustive().run(sim(|| {
        async {
            // Create an Arc-wrapped mutex
            let mutex = Arc::new(Mutex::new(5));

            // Use owned_lock to get an owned guard
            let mut guard = mutex.clone().lock_owned().await;

            // Test that we can access and modify the value through the guard
            assert_eq!(*guard, 5);
            *guard = 10;

            // Spawn a task that holds the owned guard across an await point
            async move {
                // Wait on the channel while holding the guard
                10.ms().sleep().await;

                // Verify the guard still works after the await
                assert_eq!(*guard, 10);
                *guard = 15;

                // Return the guard to verify the value later
                guard
            }
            .spawn();

            // Wait for the task to complete and get the guard back
            let guard = mutex.lock().await;

            // Verify the value was changed in the task
            assert_eq!(*guard, 15);
        }
        .primary()
        .spawn();
    }));
}

#[test]
#[should_panic = "Runtime stalled"]
fn mutex_deadlock_detection() {
    // Set up a scenario with two tasks and two mutexes where a deadlock can occur
    sim(|| {
        // Create two mutexes
        let mutex1 = Arc::new(Mutex::new(0));
        let mutex2 = Arc::new(Mutex::new(0));

        // Task 1: acquire mutex1, then try to acquire mutex2
        {
            let mutex1 = mutex1.clone();
            let mutex2 = mutex2.clone();
            async move {
                // Acquire the first mutex
                let guard1 = mutex1.lock().await;

                // Try to acquire the second mutex - this should lead to a deadlock
                // if task 2 is already holding mutex2
                let _guard2 = mutex2.lock().await;

                // Release both mutexes
                drop(_guard2);
                drop(guard1);
            }
            .primary()
            .spawn_named("Task 1");
        }

        // Task 2: acquire mutex2, then try to acquire mutex1
        {
            let mutex1 = mutex1.clone();
            let mutex2 = mutex2.clone();
            async move {
                // Acquire the second mutex
                let guard2 = mutex2.lock().await;

                // Try to acquire the first mutex - this should lead to a deadlock
                // if task 1 is already holding mutex1
                let _guard1 = mutex1.lock().await;

                // Release both mutexes
                drop(_guard1);
                drop(guard2);
            }
            .primary()
            .spawn_named("Task 2");
        }
    })();
}
