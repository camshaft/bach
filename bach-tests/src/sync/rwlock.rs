use super::*;
use crate::testing::Log;
use bach::sync::RwLock;

#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
enum Event {
    Start,
    ReadAcquired { task: usize },
    ReadReleased { task: usize },
    WriteAcquired { task: usize },
    WriteReleased { task: usize },
}

impl crate::testing::Event for Event {
    fn is_start(&self) -> bool {
        matches!(self, Self::Start)
    }
}

#[test]
fn rwlock_readers() {
    bolero::check!().exhaustive().run(sim(|| {
        async {
            let rwlock = RwLock::new(5);

            // Test multiple readers can access concurrently
            let read1 = rwlock.read().await;
            assert_eq!(*read1, 5);

            // Second reader can acquire while first reader holds the lock
            let read2 = rwlock.read().await;
            assert_eq!(*read2, 5);

            // Release first reader
            drop(read1);

            // Value should still be accessible through second reader
            assert_eq!(*read2, 5);

            // Release second reader
            drop(read2);
        }
        .primary()
        .spawn();
    }));
}

#[test]
fn rwlock_try_read() {
    bolero::check!().exhaustive().run(sim(|| {
        async {
            let rwlock = RwLock::new(10);

            // Successfully try_read
            let read1 = rwlock.try_read().unwrap();
            assert_eq!(*read1, 10);

            // Multiple readers should succeed
            let read2 = rwlock.try_read().unwrap();
            assert_eq!(*read2, 10);

            // Clean up
            drop(read1);
            drop(read2);
        }
        .primary()
        .spawn();
    }));
}

#[test]
fn rwlock_writer() {
    bolero::check!().exhaustive().run(sim(|| {
        async {
            let rwlock = RwLock::new(5);

            // Acquire write lock
            let mut write = rwlock.write().await;

            // Modify the value
            *write = 10;
            assert_eq!(*write, 10);

            // Release write lock
            drop(write);

            // Confirm value was changed
            let read = rwlock.read().await;
            assert_eq!(*read, 10);
        }
        .primary()
        .spawn();
    }));
}

#[test]
fn rwlock_try_write() {
    bolero::check!().exhaustive().run(sim(|| {
        async {
            let rwlock = RwLock::new(5);

            // Successfully try_write
            let mut write = rwlock.try_write().unwrap();
            *write = 15;
            assert_eq!(*write, 15);

            // Clean up
            drop(write);
        }
        .primary()
        .spawn();
    }));
}

#[test]
fn rwlock_read_write_exclusion() {
    bolero::check!().exhaustive().run(sim(|| {
        async {
            let rwlock = RwLock::new(5);

            // Acquire read lock
            let read = rwlock.read().await;

            // Try to write should fail while reading
            assert!(rwlock.try_write().is_err());

            // Release read lock
            drop(read);

            // Now write should succeed
            let mut write = rwlock.try_write().unwrap();
            *write = 20;

            // Try to read should fail while writing
            assert!(rwlock.try_read().is_err());

            // Release write lock
            drop(write);

            // Now read should succeed
            let read = rwlock.try_read().unwrap();
            assert_eq!(*read, 20);
        }
        .primary()
        .spawn();
    }));
}

#[test]
fn rwlock_multiple_tasks() {
    static LOG: Log<Event> = Log::new();

    bolero::check!().exhaustive().run(sim(|| {
        // Start a new event sequence
        LOG.push(Event::Start);

        let rwlock = Arc::new(RwLock::new(0));

        // Multiple reader tasks
        for i in 0..3 {
            let rwlock = rwlock.clone();
            async move {
                let read = rwlock.read().await;
                LOG.push(Event::ReadAcquired { task: i }); // Log that reader acquired lock
                bach::time::sleep(std::time::Duration::from_millis(5)).await;
                drop(read);
                LOG.push(Event::ReadReleased { task: i });
            }
            .primary()
            .spawn_named(format!("Reader {i}"));
        }

        // Writer task
        {
            let rwlock = rwlock.clone();
            async move {
                let mut write = rwlock.write().await;
                LOG.push(Event::WriteAcquired { task: 0 }); // Writer acquired lock
                *write = 50;
                bach::time::sleep(std::time::Duration::from_millis(5)).await;
                drop(write);
            }
            .primary()
            .spawn_named("Writer");
        }

        // Final reader to verify the updated value
        {
            let rwlock = rwlock.clone();
            async move {
                bach::time::sleep(std::time::Duration::from_millis(50)).await;

                let read = rwlock.read().await;
                assert_eq!(*read, 50);
            }
            .primary()
            .spawn_named("Final reader");
        }
    }));

    insta::assert_debug_snapshot!(LOG.check());
}

#[test]
fn rwlock_interleavings_snapshot() {
    static LOG: Log<Event> = Log::new();

    bolero::check!().exhaustive().run(sim(|| {
        // Start a new event sequence
        LOG.push(Event::Start);

        // Create a rwlock
        let rwlock = Arc::new(RwLock::new(0));

        // Launch two readers
        for task_id in 0..2 {
            let rwlock = rwlock.clone();
            async move {
                // Acquire read lock
                let read = rwlock.read().await;
                LOG.push(Event::ReadAcquired { task: task_id });

                // Hold the lock briefly
                bach::time::sleep(std::time::Duration::from_millis(5)).await;

                // Release the read lock
                drop(read);
                LOG.push(Event::ReadReleased { task: task_id });
            }
            .primary()
            .spawn_named(format!("Reader {task_id}"));
        }

        // Launch a writer
        {
            let rwlock = rwlock.clone();
            async move {
                // Acquire write lock
                let mut write = rwlock.write().await;
                LOG.push(Event::WriteAcquired { task: 2 });

                // Modify the value
                *write += 1;
                bach::time::sleep(std::time::Duration::from_millis(5)).await;

                // Release the write lock
                drop(write);
                LOG.push(Event::WriteReleased { task: 2 });
            }
            .primary()
            .spawn_named("Writer");
        }
    }));

    insta::assert_debug_snapshot!(LOG.check());
}

#[test]
#[should_panic = "Runtime stalled"]
fn rwlock_deadlock_detection() {
    sim(|| {
        // Create two rwlocks
        let rwlock1 = Arc::new(RwLock::new(0));
        let rwlock2 = Arc::new(RwLock::new(0));

        // Task 1: acquire write on rwlock1, then try to acquire write on rwlock2
        {
            let rwlock1 = rwlock1.clone();
            let rwlock2 = rwlock2.clone();
            async move {
                // Acquire write on first rwlock
                let mut write1 = rwlock1.write().await;
                *write1 += 1;

                // Try to acquire write on second rwlock - potential deadlock
                let mut write2 = rwlock2.write().await;
                *write2 += 1;

                drop(write2);
                drop(write1);
            }
            .primary()
            .spawn_named("Task 1");
        }

        // Task 2: acquire write on rwlock2, then try to acquire write on rwlock1
        {
            let rwlock1 = rwlock1.clone();
            let rwlock2 = rwlock2.clone();
            async move {
                // Acquire write on second rwlock
                let mut write2 = rwlock2.write().await;
                *write2 += 1;

                // Try to acquire write on first rwlock - potential deadlock
                let mut write1 = rwlock1.write().await;
                *write1 += 1;

                drop(write1);
                drop(write2);
            }
            .primary()
            .spawn_named("Task 2");
        }
    })();
}
