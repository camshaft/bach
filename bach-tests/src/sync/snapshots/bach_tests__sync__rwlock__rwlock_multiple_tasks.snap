---
source: bach-tests/src/sync/rwlock.rs
expression: LOG.check()
---
[
    Start,
    ReadAcquired {
        task: 0,
    },
    ReadAcquired {
        task: 1,
    },
    ReadAcquired {
        task: 2,
    },
    ReadReleased {
        task: 0,
    },
    ReadReleased {
        task: 1,
    },
    ReadReleased {
        task: 2,
    },
    WriteAcquired {
        task: 0,
    },
    Start,
    ReadAcquired {
        task: 0,
    },
    ReadAcquired {
        task: 1,
    },
    ReadReleased {
        task: 0,
    },
    ReadReleased {
        task: 1,
    },
    WriteAcquired {
        task: 0,
    },
    ReadAcquired {
        task: 2,
    },
    ReadReleased {
        task: 2,
    },
    Start,
    ReadAcquired {
        task: 0,
    },
    ReadAcquired {
        task: 2,
    },
    ReadAcquired {
        task: 1,
    },
    ReadReleased {
        task: 0,
    },
    ReadReleased {
        task: 2,
    },
    ReadReleased {
        task: 1,
    },
    WriteAcquired {
        task: 0,
    },
    Start,
    ReadAcquired {
        task: 0,
    },
    ReadAcquired {
        task: 2,
    },
    ReadReleased {
        task: 0,
    },
    ReadReleased {
        task: 2,
    },
    WriteAcquired {
        task: 0,
    },
    ReadAcquired {
        task: 1,
    },
    ReadReleased {
        task: 1,
    },
    Start,
    ReadAcquired {
        task: 0,
    },
    ReadReleased {
        task: 0,
    },
    WriteAcquired {
        task: 0,
    },
    ReadAcquired {
        task: 2,
    },
    ReadAcquired {
        task: 1,
    },
    ReadReleased {
        task: 2,
    },
    ReadReleased {
        task: 1,
    },
    Start,
    ReadAcquired {
        task: 0,
    },
    ReadReleased {
        task: 0,
    },
    WriteAcquired {
        task: 0,
    },
    ReadAcquired {
        task: 1,
    },
    ReadAcquired {
        task: 2,
    },
    ReadReleased {
        task: 1,
    },
    ReadReleased {
        task: 2,
    },
    Start,
    ReadAcquired {
        task: 1,
    },
    ReadAcquired {
        task: 0,
    },
    ReadAcquired {
        task: 2,
    },
    ReadReleased {
        task: 1,
    },
    ReadReleased {
        task: 0,
    },
    ReadReleased {
        task: 2,
    },
    WriteAcquired {
        task: 0,
    },
    Start,
    ReadAcquired {
        task: 1,
    },
    ReadAcquired {
        task: 0,
    },
    ReadReleased {
        task: 1,
    },
    ReadReleased {
        task: 0,
    },
    WriteAcquired {
        task: 0,
    },
    ReadAcquired {
        task: 2,
    },
    ReadReleased {
        task: 2,
    },
    Start,
    ReadAcquired {
        task: 1,
    },
    ReadAcquired {
        task: 2,
    },
    ReadAcquired {
        task: 0,
    },
    ReadReleased {
        task: 1,
    },
    ReadReleased {
        task: 2,
    },
    ReadReleased {
        task: 0,
    },
    WriteAcquired {
        task: 0,
    },
    Start,
    ReadAcquired {
        task: 1,
    },
    ReadAcquired {
        task: 2,
    },
    ReadReleased {
        task: 1,
    },
    ReadReleased {
        task: 2,
    },
    WriteAcquired {
        task: 0,
    },
    ReadAcquired {
        task: 0,
    },
    ReadReleased {
        task: 0,
    },
    Start,
    ReadAcquired {
        task: 1,
    },
    ReadReleased {
        task: 1,
    },
    WriteAcquired {
        task: 0,
    },
    ReadAcquired {
        task: 2,
    },
    ReadAcquired {
        task: 0,
    },
    ReadReleased {
        task: 2,
    },
    ReadReleased {
        task: 0,
    },
    Start,
    ReadAcquired {
        task: 1,
    },
    ReadReleased {
        task: 1,
    },
    WriteAcquired {
        task: 0,
    },
    ReadAcquired {
        task: 0,
    },
    ReadAcquired {
        task: 2,
    },
    ReadReleased {
        task: 0,
    },
    ReadReleased {
        task: 2,
    },
    Start,
    ReadAcquired {
        task: 2,
    },
    ReadAcquired {
        task: 1,
    },
    ReadAcquired {
        task: 0,
    },
    ReadReleased {
        task: 2,
    },
    ReadReleased {
        task: 1,
    },
    ReadReleased {
        task: 0,
    },
    WriteAcquired {
        task: 0,
    },
    Start,
    ReadAcquired {
        task: 2,
    },
    ReadAcquired {
        task: 1,
    },
    ReadReleased {
        task: 2,
    },
    ReadReleased {
        task: 1,
    },
    WriteAcquired {
        task: 0,
    },
    ReadAcquired {
        task: 0,
    },
    ReadReleased {
        task: 0,
    },
    Start,
    ReadAcquired {
        task: 2,
    },
    ReadAcquired {
        task: 0,
    },
    ReadAcquired {
        task: 1,
    },
    ReadReleased {
        task: 2,
    },
    ReadReleased {
        task: 0,
    },
    ReadReleased {
        task: 1,
    },
    WriteAcquired {
        task: 0,
    },
    Start,
    ReadAcquired {
        task: 2,
    },
    ReadAcquired {
        task: 0,
    },
    ReadReleased {
        task: 2,
    },
    ReadReleased {
        task: 0,
    },
    WriteAcquired {
        task: 0,
    },
    ReadAcquired {
        task: 1,
    },
    ReadReleased {
        task: 1,
    },
    Start,
    ReadAcquired {
        task: 2,
    },
    ReadReleased {
        task: 2,
    },
    WriteAcquired {
        task: 0,
    },
    ReadAcquired {
        task: 0,
    },
    ReadAcquired {
        task: 1,
    },
    ReadReleased {
        task: 0,
    },
    ReadReleased {
        task: 1,
    },
    Start,
    ReadAcquired {
        task: 2,
    },
    ReadReleased {
        task: 2,
    },
    WriteAcquired {
        task: 0,
    },
    ReadAcquired {
        task: 1,
    },
    ReadAcquired {
        task: 0,
    },
    ReadReleased {
        task: 1,
    },
    ReadReleased {
        task: 0,
    },
    Start,
    WriteAcquired {
        task: 0,
    },
    ReadAcquired {
        task: 1,
    },
    ReadAcquired {
        task: 2,
    },
    ReadAcquired {
        task: 0,
    },
    ReadReleased {
        task: 1,
    },
    ReadReleased {
        task: 2,
    },
    ReadReleased {
        task: 0,
    },
    Start,
    WriteAcquired {
        task: 0,
    },
    ReadAcquired {
        task: 1,
    },
    ReadAcquired {
        task: 0,
    },
    ReadAcquired {
        task: 2,
    },
    ReadReleased {
        task: 1,
    },
    ReadReleased {
        task: 0,
    },
    ReadReleased {
        task: 2,
    },
    Start,
    WriteAcquired {
        task: 0,
    },
    ReadAcquired {
        task: 2,
    },
    ReadAcquired {
        task: 1,
    },
    ReadAcquired {
        task: 0,
    },
    ReadReleased {
        task: 2,
    },
    ReadReleased {
        task: 1,
    },
    ReadReleased {
        task: 0,
    },
    Start,
    WriteAcquired {
        task: 0,
    },
    ReadAcquired {
        task: 2,
    },
    ReadAcquired {
        task: 0,
    },
    ReadAcquired {
        task: 1,
    },
    ReadReleased {
        task: 2,
    },
    ReadReleased {
        task: 0,
    },
    ReadReleased {
        task: 1,
    },
    Start,
    WriteAcquired {
        task: 0,
    },
    ReadAcquired {
        task: 0,
    },
    ReadAcquired {
        task: 2,
    },
    ReadAcquired {
        task: 1,
    },
    ReadReleased {
        task: 0,
    },
    ReadReleased {
        task: 2,
    },
    ReadReleased {
        task: 1,
    },
    Start,
    WriteAcquired {
        task: 0,
    },
    ReadAcquired {
        task: 0,
    },
    ReadAcquired {
        task: 1,
    },
    ReadAcquired {
        task: 2,
    },
    ReadReleased {
        task: 0,
    },
    ReadReleased {
        task: 1,
    },
    ReadReleased {
        task: 2,
    },
]
